DOCS-8058 - SAP S/4HANA OData cloud connector 2.1.0 (#757)
mulesoft/docs-connectors: modules/ROOT/pages/sap/sap-s4hana-cloud-connector-create-batch.adoc

= Batch Processing
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

The batch processing feature is available starting in SAP S/4HANA OData Cloud Connector version 2.1. 

> You must follow the correct order for structural batch operations:

Reviewer: If "you" must follow the correct order, then I think the list items should also be worded for "you".

Otherwise, you might want a differently worded introduction, such as "A structural batch operation must follow a particular order to be successful:"

Also, does "batch processing" comprise "structural batch operations"; is this different from "batch operations"?

 
. The batch must be initiated before any requests are added into it.

> . Requests cannot be added to the batch or the change set after it is executed.

Reviewer: Will your audience know what a "change set" is and why you equated it with the batch here?

Reviewer 2: The change set is documented in the reference, which is linked in See Also.

== Create a Batch Operation

To create and process a batch operation, you must:

. Place the *Initiate Batch* operation in the Mule flow, which returns a unique batch identifier. +

> This unique batch identifier should be stored, for example, in the flow variable, so that it can then be used to group the batch data operations under the common batch request.

Reviewer: ```suggestion
You can use store this unique batch identifier in the flow variable, for example, to group batch data operations in a common batch request.
```
OR
```suggestion
One use for this identifier, for example, is to group batch data operations in a common batch request by storing the identifier in the flow variable.
```

. Use the *Create Change Set* operation to create a new change set. +

> This operation returns a unique change set identifier, which must then be used in the *Batch Create*, *Batch Update*, *Delete Entity*, or *Batch Execute Function* operations in order to add them to the change set.

Reviewer: ```suggestion
This operation returns a unique change set identifier, which you then use in the *Batch Create*, *Batch Update*, *Delete Entity*, or *Batch Execute Function* operations to add them to the change set.
```

+
The *Batch Create Entity*, *Batch Get Entity*, *Batch Query*, and *Batch Execute Function* operations return a JSON result, however, the result of these operations cannot be read until the *Execute Batch* operation is executed. 
+

> If you attempt to read the result before the *Execute Batch* operation is executed, the connector throws an exception. You can store the output of these operations, for instance, into a flow variable, and access the output after the *Execute Batch* operation is executed.

Reviewer: This is confusing.

The second **step** to creating and processing a batch operation is to create a change set.

But we then follow it with a bunch of other instructions about how to use the **result** of that step:

The result of the step is a unique ID. 
...that you must then use (more steps?) in four operations (call them A, B, C, and D) to add them to the created change set.

But then we mention that three completely different operations (E, F, and G) and only one of the previous four operations (D) return a result (from being added to the created change set) that can't be read until a fourth different operation (H) is executed (another step?) and if you try, you get an error.

And then we say that you can store (another step?) the result of E, F, G, and D (for example, in a flow variable) and then access (another step?) that result after H is executed.

And all of this following the result of creating the change set is the actual batch process referred to as "complete" in the next real step?

I feel like I need a graphic to keep this straight. :-)

Reviewer: After reading further, it seems to me that Lines 22-24 actually should go AFTER the following step on Line 26.

+

> . When the batch is complete, use the *Execute Batch* operation to send a batch request.

Reviewer: ```suggestion
. When the batch finishes, use the *Execute Batch* operation to send a batch request.
```
So what actions make up "the batch" (operation): all of the previous text?
And the result of using *Execute Batch* (H) means that you can access the stored JSON results of E, F, G, and D?

Reviewer 2: In this case, the batch isn't doing anything. You are adding different operations into a batch (so complete is the right word)

> . After the *Execute Batch* operation is successfully executed, a summary is returned, which contains basic information about the outcome of every batch operation that was included as part of the batch operation. The detailed results of these batch operations are then available to read.

Reviewer: This is not a step; there is no user action here, except maybe navigating to and viewing "the detailed results of these batch operations."

Also, isn't the result of the *EB* operation the batch request, and isn't it the batch request that "is successfully executed"?

And this text doesn't make sense: "every batch operation (that was included as part) of the batch operation." 
The parts of the batch operation are batch operations?

I made a guess:

```suggestion
After the batch request is successfully executed, a summary of every batch operation that was included as part of the batch request is returned. You can then read detailed results of these batch operations <where? how?>.
```
If the summary "is returned," how does the audience get to "detailed results"?

== Troubleshooting

Exceptions are thrown in the following circumstances:

> * The connector throws an exception if an operation references an invalid batch identifier, either because it was never initiated or if the *Execute Batch* operation was already performed for that particular batch identifier. 

Reviewer: ```suggestion
* An operation references a batch identifier that is invalid, either because:
** It was never initiated.
** It was already processed as part of an *Execute Batch* operation. 
```

> * An exception is thrown when the change set identifier used in that batch operation is not valid.

Reviewer: ```suggestion
* When the change set identifier is not valid.
```
Would it be correct to avoid the details and say simply that exceptions are thrown if either a batch identifier or a change set identifier is not valid?
